controlplane/kubeadm/internal/failure_domain.go

```diff
@@ -50,15 +50,32 @@ func (f failureDomainAggregations) Swap(i, j int) {
 	f[i], f[j] = f[j], f[i]
 }
 
-// PickMost returns the failure domain with the most number of machines.
-func PickMost(failureDomains clusterv1.FailureDomains, machines FilterableMachineCollection) *string {
+// PickMost picks a machine among machines that has the failure domain with most control plane machines on.
+// After getting an ordered failure domain list based on number of control plane machines, return the machine that matches the first failure domain in the list.
+func PickMost(c *ControlPlane, machines FilterableMachineCollection) *string {
+	// orderDescending sorts failure domains according to all  control plane machines
+	fds := orderDescending(c.Cluster.Status.FailureDomains.FilterControlPlane(), c.Machines)
+	for _, fd := range fds {
+		for _, m := range machines {
+			if m.Spec.FailureDomain == nil {
+				return &fd.id
```

Comment:
Why does it return a failure domain if the machine is not in a failure domain?

Reply:
You are right, it should continue in this case.